go/kbfs/libkbfs/prefetcher.go

```diff
@@ -719,6 +750,8 @@ func (p *blockPrefetcher) prefetchDirectDirBlock(
 		numBlocks += b
 		numBytesFetched += f
 		numBytesTotal += t
+
+		p.handleCriticalRequests()
```

Comment:
Why are these all in the individual handlers instead of just before the return in `handlePrefetch`, or just after `handlePrefetch` is called in the `run` loop?

Reply:
`handlePrefetch()` isn't called in a loop, it's just called once per request.  Do you mean `request()`?  I guess I just liked being more explicit about it, and `request()` didn't seem like a natural place for it; I guess I was worried that `request()` was too generic and could be reasonably called from elsewhere in the code.   If you want I'm happy to change it though.

Reply:
`prefetch{Indirect,Direct}{Dir,File}Block()` are called in `handlePrefetch()`, which is in turn called in the `run()` loop. If you put a single `handleCriticalRequests` call in either of `handlePrefetch()` or `run()` after the call to `handlePrefetch`, you can replace these three `handleCriticalRequests()` calls with one.

Reply:
The point of this is that it's called inside the `for` loop inside those `prefetch{Indirect,Direct}{Dir,File}Block()` calls.  It can't just be a single call in `run()` (in fact, there already is one there).

Reply:
Oh I see. What's the benefit here though? The calls inside the for loop are all async, we should be able to preempt them with a high-priority critical request.

Reply:
The call isn't completely async -- it does a `checkCaches` call inline.  Add 2000 of those up, and you're talking about 10s of seconds inside that for loop.

Reply:
(because of the leveldb calls)

Reply:
Hmm maybe we should eliminate the disk cache check in `checkCaches`, allowing the primary queue in `block_retrieval_queue` to do that as part of the call out to `bserver_remote`.

Reply:
Maybe but that seems like a larger discussion as it would slow down normal block accesses a bit by forcing them to go through the normal queue when they're already cached.

Reply:
OK I agree it's a larger discussion, but it seems like the right call. I don't love this critical request handling in the depths of a prefetch request being handled. The current design necessitates it, which is a strong signal for me that we should change the current design.

Reply:
Sounds good, feel free to open a ticket.  We could also just add a signal to `blockRetrievalQueue.Request()` to skip the disk cache check just for this given request, and use that in `prefetcher.request()`.